os/arch/arm/src/amebad/amebad_i2c.c

```diff
@@ -0,0 +1,1237 @@
+static inline void amebad_i2c_sem_destroy(FAR struct amebad_i2c_priv_s *priv);
+
+#ifdef CONFIG_I2C_TRACE
+static void amebad_i2c_tracereset(FAR struct amebad_i2c_priv_s *priv);
+static void amebad_i2c_tracenew(FAR struct amebad_i2c_priv_s *priv, uint32_t status);
+static void amebad_i2c_traceevent(FAR struct amebad_i2c_priv_s *priv, enum amebad_trace_e event, uint32_t parm);
+static void amebad_i2c_tracedump(FAR struct amebad_i2c_priv_s *priv);
+#endif							/* CONFIG_I2C_TRACE */
+
+static void amebad_i2c_setclock(FAR struct amebad_i2c_priv_s *priv, uint32_t frequency);
+static inline void amebad_i2c_sendstart(FAR struct amebad_i2c_priv_s *priv, uint8_t address);
+static inline void amebad_i2c_sendstop(FAR struct amebad_i2c_priv_s *priv);
+static inline uint32_t amebad_i2c_getstatus(FAR struct amebad_i2c_priv_s *priv);
+
+static int amebad_i2c_isr_process(struct amebad_i2c_priv_s *priv);
+
+#ifndef CONFIG_I2C_POLLED
+static int amebad_i2c_isr(int irq, void *context, FAR void *arg);
+#endif							/* !CONFIG_I2C_POLLED */
+
+static int amebad_i2c_init(FAR struct amebad_i2c_priv_s *priv);
+static int amebad_i2c_deinit(FAR struct amebad_i2c_priv_s *priv);
+#ifdef CONFIG_I2C_TRANSFER
+static int amebad_i2c_transfer(FAR struct i2c_dev_s *dev, FAR struct i2c_msg_s *msgs, int count);
+#endif
+#ifdef CONFIG_I2C_RESET
+static int amebad_i2c_reset(FAR struct i2c_master_s *dev);
+#endif
+
+static int amebad_i2c_process(FAR struct i2c_dev_s *dev, FAR struct i2c_msg_s *msgs, int count);
+static uint32_t amebad_i2c_setfrequency(FAR struct i2c_dev_s *dev, uint32_t frequency);
+static int amebad_i2c_setaddress(FAR struct i2c_dev_s *dev, int addr, int nbits);
+static int amebad_i2c_write(FAR struct i2c_dev_s *dev, const uint8_t *buffer, int buflen);
+static int amebad_i2c_read(FAR struct i2c_dev_s *dev, uint8_t *buffer, int buflen);
+#ifdef CONFIG_I2C_WRITEREAD
+static int amebad_i2c_writeread(FAR struct i2c_dev_s *dev, const uint8_t *wbuffer, int wbuflen, uint8_t *buffer, int buflen);
+#endif
+
+/************************************************************************************
+ * Private Data
+ ************************************************************************************/
+
+/* Trace events strings */
+
+#ifdef CONFIG_I2C_TRACE
+static const char *g_trace_names[] = {
+	"NONE           ",
+	"SENDADDR       ",
+	"SENDBYTE       ",
+	"RCVBYTE        ",
+	"NOSTART        ",
+	"START/RESTART  ",
+	"STOP           ",
+	"ERROR          "
+};
+#endif
+
+/* I2C interface */
+
+static const struct i2c_ops_s amebad_i2c_ops = {
+	.setfrequency = amebad_i2c_setfrequency,
+	.setaddress = amebad_i2c_setaddress,
+	.write = amebad_i2c_write,
+	.read = amebad_i2c_read,
+#ifdef CONFIG_I2C_WRITEREAD
+	.writeread = amebad_i2c_writeread,
+#endif
+#ifdef CONFIG_I2C_TRANSFER
+	.transfer = amebad_i2c_transfer,
+#endif
+#ifdef CONFIG_I2C_SLAVE
+	.setownaddress = amebad_i2c_setaddress, //amebad_i2c_setownaddress,
+	//.registercallback = amebad_i2c_registercallback
```

Comment:
if we enable this config, this will occur error, please enable this.

Reply:
Done.